containers-tests/tests/intmap-strictness.hs

```diff
@@ -68,12 +70,35 @@ prop_strictFromSet :: Func Key (Bot A) -> IntSet -> Property
 prop_strictFromSet fun set =
   isBottom (M.fromSet f set) === any (isBottom . f) (IntSet.toList set)
   where
-    f = coerce (applyFunc fun) :: Key -> A
+    f = applyFunc fun
+
+prop_strictFromSetA :: Func Key (Bot A) -> IntSet -> Property
+prop_strictFromSetA fun set =
+  isBottom (getSolo (M.fromSetA (MkSolo . f) set)) === any (isBottom . f) (IntSet.toList set)
+  where
+    f = applyFunc fun
 
 prop_lazyFromSet :: Func Key (Bot A) -> IntSet -> Property
 prop_lazyFromSet fun set = isNotBottomProp (L.fromSet f set)
   where
-    f = coerce (applyFunc fun) :: Key -> A
+    f = applyFunc fun
+
+prop_lazyFromSetA :: Func Key (Bot A) -> IntSet -> Property
+prop_lazyFromSetA fun set = isNotBottomProp (getSolo (L.fromSetA f set))
+  where
+  f = MkSolo . applyFunc fun
+
+prop_fromSetA_equiv_strictness :: Func Int (Bot A) -> IntSet -> Property
+prop_fromSetA_equiv_strictness fun set =
+  -- strict fromSetA is the same as lazy and then forcing
+  bottomOn (M.fromSetA f set) (fmap forceValues (L.fromSetA f set)) .&&.
+  -- strict fromSetA is the same as lazy fromSetA composed with strictly applied
+  -- wrapper
+  bottomOn (M.fromSetA f set) (fmap getSolo . getCompose $ L.fromSetA (Compose . fmap (MkSolo $!) . f) set)
+  where
+  forceValues xs = foldr (\ !_ r -> r) () xs `seq` xs
+  bottomOn = on (===) (isBottom . getSolo)
```

Comment:
```suggestion
  bottomOn = (===) `on` (isBottom . getSolo)
```

<strike>The absolute objective reality</strike> My very strong opinion is that `on` should only ever be used as an infix operator.